### ISSUE TITLE
Very slow loading in version 2.3.0

### ISSUE THREAD

**Issue by sunbyte** [NanoHttpd/nanohttpd#318]
Hello, after update from version 2.1.1 to the last version 2.3.0 of the library (with some code modifications), every request to the server (loading simple HTML for example) is executed very slowly and load up to ~5 seconds each. When I switch back to version 2.1.1 of the library everything is okay.

**Comment by alexruskovski** [comment]
sunbyte, did you find a solution. Im facing with the same problem now. Please if you found a way post your solution. thanks

**Comment by LordFokas** [comment]
Can you guys checkout the current in-dev version (3.0.0) and see if it still happens?

**Comment by alexruskovski** [comment]
Thank you for your quick response. I downgrade my NanoHTTPD to 2.1.1 version and now is working as it should be. Thanks

**Comment by chenyayun** [comment]
I use 2.3.1 version, and load up to 10 seconds. Is there any solutions?

**Comment by chenyayun** [comment]
Hey guys! I found the result. 

In HTTPSession constructor method, 

>     public HTTPSession(NanoHTTPD httpd, ITempFileManager tempFileManager, InputStream inputStream, OutputStream outputStream, InetAddress inetAddress) 
{
        this.httpd = httpd;
        this.tempFileManager = tempFileManager;
        this.inputStream = new BufferedInputStream(inputStream, HTTPSession.BUFSIZE);
        this.outputStream = outputStream;
        this.remoteIp = inetAddress.isLoopbackAddress() || inetAddress.isAnyLocalAddress() ? "127.0.0.1" : inetAddress.getHostAddress().toString();
        this.remoteHostname = inetAddress.isLoopbackAddress() || inetAddress.isAnyLocalAddress() ? "localhost" : inetAddress.getHostName().toString();
        this.headers = new HashMap<String, String>();
    }

inetAddress.getHostName()  consume to much time. I comment out these code, and it work fine, is very fast to load.

**Comment by 835919230** [comment]
Why inetAddress.getHostName() take to much time?

**Comment by LordFokas** [comment]
Because it has to resolve the domain / host name of the client's IP, which can require several lookups across the network / internet. I've already developed a fix in v3.0.0 for this, once it is deployed those kind of problems will cease to exist.

**Comment by teras** [comment]
I tried the latest git version, and I still have issues when uploading a file.
Specifically there is a 3 second (constant) delay to read the first byte every time.

The test has been performed locally and there shouldn't be an issue with the hostname, which I disabled and still has a problem.

Is this the same problem or something else?

**Comment by LordFokas** [comment]
Nah, this should be something else. Do you have the tools to profile the cause?

**Comment by teras** [comment]
No, I have created a simple consume method, which did the following (it's in Kotlin, sorry, but you get the idea):

    val size = Integer.parseInt(session?.headers?.get("content-length") ?: return invalidUpload())
    val out = ByteArrayOutputStream(size)
    val buffer = ByteArray(Math.min(0x8000, size))
    val stream = BufferedInputStream(session.inputStream)
    println("Start")
    val time = System.currentTimeMillis()
    out.write(stream.read())
    println("<dt> = ${(System.currentTimeMillis() - time) / 1000f}")

and invoke it with curl like this:

    curl -X PUT --data-binary "@target/nanohttpd-2.3.2-SNAPSHOT.jar" --user a:user http://localhost:8080/test

**Comment by balrok** [comment]
I can confirm this for a simple POST method.
On windows it needs ~15s until my serve-method will even start.. on Linux everything works fine.
(Maybe this is not about Win/Linux but about the network: my Linux PC is on the internet, my Windows PC can not access the internet)

**Comment by trnchawla** [comment]
> Can you guys checkout the current in-dev version (3.0.0) and see if it still happens?

Where to get this branch, am not able to find it.

**Comment by hervegirod** [comment]
It is slow in some configurations because of the `inetAddress.getHostAddress(`) in the `HTTPSession `code (same code in 2.3.0 and current version). You should rather use `inetAddress.toString()` and get rid of the "/" and the content after it, it will be much quicker.
